client/src/Try/Editor.purs

```diff
@@ -0,0 +1,234 @@
+import Web.HTML (HTMLElement)
+
+type Slot = H.Slot Query Output
+
+data Query a
+  = GetEditorContent (Maybe String -> a)
+  | SetEditorContent String a
+  | SetAnnotations (Array Annotation) a
+  | AddMarker MarkerType ErrorPosition a
+  | RemoveMarkers a
+
+data Output = TextChanged String
+
+data MarkerType = MarkerError | MarkerWarning
+
+toStringMarkerType :: MarkerType -> String
+toStringMarkerType = case _ of
+  MarkerError -> "error"
+  MarkerWarning -> "warning"
+
+newtype MarkerId = MarkerId Int
+derive instance newtypeMarkerId :: Newtype MarkerId _
+
+derive instance eqMarkerType :: Eq MarkerType
+
+type Debouncer =
+  { var :: AVar Unit
+  , fiber :: Fiber Unit
+  }
+
+debounceTime :: Milliseconds
+debounceTime = Milliseconds 750.0
+
+type State =
+  { editor :: Maybe Editor
+  , markers :: List MarkerId
+  , debounceRef :: Maybe (Ref (Maybe Debouncer))
+  }
+
+data Action
+  = Initialize
+  | Finalize
+  | ClearMarkers
+  | HandleChange
+
+component :: forall i m. MonadAff m => H.Component HH.HTML Query i Output m
+component = H.mkComponent
+  { initialState
+  , render
+  , eval: H.mkEval $ H.defaultEval
+      { handleAction = handleAction
+      , handleQuery = handleQuery
+      , initialize = Just Initialize
+      , finalize = Just Finalize
+      }
+  }
+  where
+  initialState :: i -> State
+  initialState _ =
+    { editor: Nothing
+    , markers: Nil
+    , debounceRef: Nothing
+    }
+
+  -- As we're embedding a 3rd party component we only need to create a placeholder
+  -- div here and attach the ref property which will let us reference the element.
+  render :: State -> H.ComponentHTML Action () m
+  render _ =
+    HH.div
+      [ HP.ref $ H.RefLabel "ace"
+      , HP.id_ "code"
+      ]
+      [ ]
+
+  handleAction :: Action -> H.HalogenM State Action () Output m Unit
+  handleAction = case _ of
+    Initialize -> do
+      H.getHTMLElementRef (H.RefLabel "ace") >>= traverse_ \element -> do
+        debounceRef <- H.liftEffect $ Ref.new Nothing
+        editor <- H.liftEffect $ setupEditor element
+        H.modify_ _ { editor = Just editor, debounceRef = Just debounceRef }
+        session <- H.liftEffect $ Editor.getSession editor
+        void $ H.subscribe $ ES.effectEventSource \emitter -> do
+          EditSession.onChange session (\_ -> ES.emit emitter HandleChange)
+          pure mempty
+
+    Finalize -> do
+      handleAction ClearMarkers
+      H.modify_ _ { editor = Nothing }
+
+    ClearMarkers -> do
+      { editor: mbEditor, markers } <- H.get
+      for_ mbEditor \editor -> H.liftEffect do
+        session <- Editor.getSession editor
+        let cleanup (MarkerId n) = EditSession.removeMarker n session
+        traverse_ cleanup markers
+
+    HandleChange -> do
+      let
+        handleChange = H.gets _.editor >>= traverse_ \editor -> do
+          text <- H.liftEffect (Editor.getValue editor)
+          H.raise $ TextChanged text
+
+      H.gets _.debounceRef >>= traverse_ \ref -> do
```

Comment:
Can we extract this debouncing logic out? It's a little complex, and I think currently more mixed up with the editor component logic than it needs to be. I also think it might be nice to consider implementing it in terms of `setTimeout` rather than `AVar`, like the lodash version is.

I'm imagining something like a function `debounce :: forall a. (a -> Aff Unit) -> Milliseconds -> Aff (a -> Aff Unit)` which would be run once when the component mounts; the outer `Aff` in the return value would be to allow the debouncing function to set up the state it needs.

Reply:
It turns out this is just done in `Effect` anyways, so I could write a simple debouncer using `setTimeout` and `clearTimeout`. The other debouncer allows full evaluation in `HalogenM`, but you're right that it's overkill here. I've tested out the new implementation and it works fine.